feat: add color scheme support

edouard-lopez opened the pull request:
fixes: #8 

This is based on #10 as I updated the main dependencies to their latest version.

Color schemes follow [XTerm Itheme interface](https://xtermjs.org/docs/api/terminal/interfaces/itheme/). You can see [live example color schemes in `pure` (fish)](https://github.com/pure-fish/pure/tree/120b08d6d59e197649f7d6ddfffdc2d2f8feb4d3/tools/colorscheme).

### Usage

<table>
<tr>
<th>Command
<th>Screenshot
<tr>
 <td>
<pre>
fish_prompt \
      | node out/src/cli.js -o cli.png \
         --color-scheme ./tests/ayu.json \
         --margin 5 \
         --background-color '#0f1419'
</pre> 
 <td>
<img width="640" height="42" alt="image" src="https://github.com/user-attachments/assets/4ce52e63-bf1f-4b80-ac8a-917e5fc09f7b" />
<tr>
 <td>
<pre>
fish_prompt \
      | node out/src/cli.js -o cli.png \
         --color-scheme ./tests/ayu-light.json \
         --margin 5 \
         --background-color '#FAFAFA'
</pre> 
 <td>
<img width="640" height="42" alt="image" src="https://github.com/user-attachments/assets/77c5ebe3-01d7-4b1f-90b0-1b771c3451b8" />
</table>

edouard-lopez committed:
chore: update packages setup eslint and simplify prettier config as we use default

* puppeteer to 9x → 21.x
* typescript 4.x → 5.x
* ts-node 9.x → 10.x
* jest 26.x → 29.x
* jest-image-snapshot 4.x → 6.x

edouard-lopez committed:
feat(cli): add --color-scheme option to CLI

fixes #8

OmarTawfik reviewed (changes requested):
@edouard-lopez thank you so much for looking into this!
But same suggestion about not mixing this upgrade with other drive-by changes, that are not required/related. I see this PR includes such changes, and also changes from #10 .. I'm happy to finalize/close #10 first before spending more time on #11 if you would like.

Thanks!

edouard-lopez committed:
fix: fallback to background color when no theme provided

edouard-lopez committed:
feat: add option to set color scheme/theme

edouard-lopez commented:
I'm puting this one on draft until you merge the other one #10, #12 and #14

edouard-lopez committed:
fix: correct build error TS2403: AbortSignal' was also declared here.

$ tsc
Error: node_modules/@types/node/globals.d.ts(341,13): error TS2403: Subsequent variable declarations must have the same type.  Variable 'AbortSignal' must be of type '{ new (): AbortSignal; prototype: AbortSignal; abort(reason?: any): AbortSignal; timeout(milliseconds: number): AbortSignal; }', but here has type '{ new (): AbortSignal; prototype: AbortSignal; }'.

edouard-lopez commented:
> Good idea. Maybe we can add an options.theme to replace options.backgroundColor .. I suggest accepting the xterm's theme object ([link](https://xtermjs.org/docs/api/terminal/interfaces/iterminaloptions/#optional-theme)). I'm a bit skeptic about depending on the xterm-theme package. It is a 5 years old package with no real usage or known contributors. But the suggestion above would permit users of terminal-screenshot to leverage xterm-theme if they want, and forward its data along.
@OmarTawfik 

I change the approach to above

OmarTawfik commented on src/options.ts:
```diff
@@ -11,6 +12,8 @@ export interface TerminalScreenshotOptions {
   // Type of the screenshot to be generated. (default: png)
   readonly type: "png" | "jpeg";
+  // Name of the theme to use. (default: "")
+  readonly theme: ITheme; // values https://github.com/ysk2014/xterm-theme/blob/master/src/index.js
```
Given the breaking change, I suggest adding a `BREAKING CHANGE` snippet to the PR, to explain the API/CLI additions/removals:

https://www.npmjs.com/package/semantic-release

edouard-lopez replied:
~BREAKING CHANGE will be needed if the new version breaks previous use cases, here we are only adding a new way to customize, not removing previous behaviour IMHO. WDYT?~

OmarTawfik commented on src/cli.ts:
```diff
@@ -6,13 +6,14 @@ import {renderScreenshot} from "./index";
-  .option("-d, --data [string]", "Data to be render to the terminal.")
-  .option("-m --margin [number]", "Margin to leave around the terminal area in pixels. (default: 0)")
-  .option("-f --font-family [string]", "Font family to use in terminal output. (default: Monaco)")
   .option("-b --background-color [css-color]", "Background color of the terminal. (default: black)")
```
Given the capabilities of the new `xterm.Theme`, I wonder if we should also remove `--background-color` and `--margin`, and rely solely on the xterm theme configuration .. WDYT?

edouard-lopez replied:
For `--background-color` I agree, but there is [no `margin` property in ITheme](https://xtermjs.org/docs/api/terminal/interfaces/itheme/), so we should keep it

OmarTawfik replied:
I got some time today to experiment with it, and it looks like we have to keep both. `--background-color` is different than `ITheme.background`, as the former will be used in the `--border`, while the latter won't. So I take back my original suggestion.

OmarTawfik commented on src/cli.ts:
```diff
@@ -6,13 +6,14 @@ import {renderScreenshot} from "./index";
-  .option("-m --margin [number]", "Margin to leave around the terminal area in pixels. (default: 0)")
-  .option("-f --font-family [string]", "Font family to use in terminal output. (default: Monaco)")
   .option("-b --background-color [css-color]", "Background color of the terminal. (default: black)")
+  .option("-c --color-scheme [string]        Color scheme of the terminal, i.e. theme. (default: '')")
```
I suggest using `--theme` instead of `--color-scheme`, and accepting a path to a JSON file, containing the theme. This should give users the maximum flexibility, without us having to explicitly override/configure this object. It would also maintain compatibility with the API (`Config.theme` is an object, not a name).

edouard-lopez replied:
I agree using `theme` would be better, but we already have a `--type`/`-t` option, so using `--theme`/`-t` option is in conflict. I see several options here:
1. rename `--type` to ~`--format`/`-f`~ `--extension-type`/`-e` (we have `--font-family`/`-f`)
2. keep `--color-scheme`/`-c` <-- will stick to that for the time being

OmarTawfik replied:
Good catch. In that case, I think `-k --kind [png|jpeg]` would work better instead of `-e --extension-type`, as users might confuse that as appending an "extension" to the existing output file path.

OmarTawfik commented on src/options.ts:
```diff
@@ -11,6 +12,8 @@ export interface TerminalScreenshotOptions {
   readonly backgroundColor: string;
   // Type of the screenshot to be generated. (default: png)
```
related to the other comments, I think it is an opportunity to remove `backgroundColor` and `margin` options, and rely on the theme config. WDYT?

edouard-lopez replied:
superseded by https://github.com/OmarTawfik/terminal-screenshot/pull/11#discussion_r1492581182

OmarTawfik commented on src/options.ts:
```diff
@@ -11,6 +12,8 @@ export interface TerminalScreenshotOptions {
   readonly backgroundColor: string;
   // Type of the screenshot to be generated. (default: png)
   readonly type: "png" | "jpeg";
+  // Name of the theme to use. (default: "")
```
nit: as this is the theme definition itself, not the name of the theme:

```suggestion
  // XTerm theme overrides to use:
  // https://github.com/xtermjs/xterm.js/blob/5.3.0/typings/xterm.d.ts#L323-L375
```

OmarTawfik commented on src/options.ts:
```diff
@@ -19,4 +22,5 @@ export const TerminalScreenshotOptionsSchema = joi.object({
   fontFamily: joi.string().min(1).default("Monaco"),
   backgroundColor: joi.string().min(1).default("black"),
   type: joi.string().valid("png", "jpeg").default("png"),
+  theme: joi.object<ITheme>().min(1).default(""),
```
nit: as it defaults to `undefined`, which is a supported value for the object, rather than an empty string:

```suggestion
  theme: joi.object<ITheme>().min(1),
```

OmarTawfik replied:
In the recent version, this was changed again from `ITheme` to a `string`. Is that intended?

The API version is completely in-memory. For example, it doesn't take the output path, but returns the produced image as a buffer.
This allows doing things like running it in tests, where setup/clean up is purely done in memory.

I suggest keeping it the same here, taking the theme object `joi.object<ITheme>()` directly instead of its path.
The CLI can still take a path and load it from there of course.

OmarTawfik commented on tests/options.test.ts:
```diff
@@ -9,6 +9,8 @@ defineTest("maximum margin", {data: "test", margin: 10000000}, `"margin" must be
 
 defineTest("invalid type", {data: "test", type: "foo"}, `"type" must be one of [png, jpeg]`);
 
```
I suggest adding a test with an invalid property of `Itheme`, for example:

```suggestion

defineTest("invalid theme property", {data: "test", theme: {foreground: "black", fooBar: "invalid"}}, ...);
```

edouard-lopez replied:
won't do as we switch the option to a filepath

OmarTawfik replied:
This is an API test, not a CLI one.

OmarTawfik commented on src/options.ts:
```diff
@@ -11,6 +12,8 @@ export interface TerminalScreenshotOptions {
   readonly type: "png" | "jpeg";
+  // Name of the theme to use. (default: "")
+  readonly theme: ITheme; // values https://github.com/ysk2014/xterm-theme/blob/master/src/index.js
 }
 
```
I suggest not mentioning or depending on `xterm-theme` (here + `README` + others), as it is unmaintained, and without proper docs/policy in place. The users can always refer to the official `xterm` package if needed.

edouard-lopez replied:
I will remove the dependency but still mention `xterm-theme` as it provide some quickstart for writing a theme

OmarTawfik replied:
Perhaps we can refer to the example this PR will be adding to our `main` branch instaed of the unknown/unmaintained repo? For example:

- https://github.com/OmarTawfik/terminal-screenshot/blob/main/tests/ayu.json

edouard-lopez replied:
Tinted theming project has a lot of [schemes available](https://github.com/tinted-theming/schemes) and explains [how to write one](https://github.com/tinted-theming/home/blob/db6f4bfb839bbf3390bbf51a40341369c11f90dd/styling.md#specific-colours-and-their-usages).

edouard-lopez committed:
refactor: change to ITheme object for theme

BREAKING CHANGE:


# Conflicts:
#	tests/screenshots/background-dark.png
#	tests/screenshots/background-light.png
#	tests/screenshots/margin.png

OmarTawfik reviewed (changes requested):
Looks like this needs rebasing, given everything else is merged already.

I left a few suggestions. Thanks!

edouard-lopez committed:
chore: correct typo

edouard-lopez committed:
refactor: back to --color-scheme option and expect filepath

BREAKING CHANGE:
* from --theme to --color-scheme
* from ITheme to filepath to theme definition

edouard-lopez commented:
@OmarTawfik I rebase the PR and expect a file path option to define the color scheme. Also went back to the color scheme nomenclature to avoid cli conflict with `--type`

OmarTawfik commented on tests/screenshots.test.ts:
```diff
@@ -40,6 +40,25 @@ function dark(...parts: string[]): string {
   return colors.bgHex("#000").hex("#FFF")(parts.join(""));
 }
 
+defineTest("colorScheme: dark", {
```
I suggest:

- Using `kabab-case` in test names like existing tests for uniformity, as characters like colons `:` are not supported on other file systems like Windows.
- Putting the new tests above with the existing ones, instead of in utils here.

OmarTawfik commented on yarn.lock:
```diff
@@ -2496,9 +2496,9 @@ eastasianwidth@^0.2.0:
-  version "1.4.659"
-  resolved "https://registry.yarnpkg.com/electron-to-chromium/-/electron-to-chromium-1.4.659.tgz#e93af8119b6610cb4d2614a47508a74543b96ce5"
-  integrity sha512-sRJ3nV3HowrYpBtPF9bASQV7OW49IgZC01Xiq43WfSE3RTCkK0/JidoCmR73Hyc1mN+l/H4Yqx0eNiomvExFZg==
+  version "1.4.661"
```
nit: unintended changes?
I suggest reverting.

edouard-lopez committed:
chore: update to match feedbacks

* kebab-case
* grouping tests
* update screenshots
* option `type` -> `kind`

edouard-lopez commented:
Hello, 
long time no see. Apology, I took a one-year hiatus from coding since last time.
Hope everything is good for you!

I pushed the changes you requested. Once this is merged, I will be able to resume my PR to [automatically generate screenshots for `pure`'s documentation!](https://github.com/pure-fish/pure/pull/345) :tada:

edouard-lopez committed:
refactor: read colorScheme from JSON file

* update doc
* add 10 000ms timeout for screenshot tests

edouard-lopez commented:
Hello,
I created a [docker image](https://github.com/pure-fish/docker-terminal-screenshot/) based on this PR to [use in `pure` CI ](https://github.com/pure-fish/pure/blob/120b08d6d59e197649f7d6ddfffdc2d2f8feb4d3/.github/workflows/doc.yml#L34) for the purpose of generating [live screenshots in the project's documentation](https://github.com/pure-fish/pure/pull/386).

Feedback welcome :)
